File: sdk/packages/go/auth_test.go
package whodb

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"
)

// fakeCli writes an executable shell script standing in for the whodb CLI.
// It appends one line to countFile per invocation so tests can assert on
// exec counts (cache behavior).
func fakeCli(t *testing.T, body string) (command, countFile string) {
	t.Helper()
	dir := t.TempDir()
	countFile = filepath.Join(dir, "count")
	command = filepath.Join(dir, "whodb")
	script := fmt.Sprintf("#!/bin/sh\necho x >> %q\n%s\n", countFile, body)
	if err := os.WriteFile(command, []byte(script), 0o755); err != nil {
		t.Fatal(err)
	}
	return command, countFile
}

func execCount(t *testing.T, countFile string) int {
	t.Helper()
	data, err := os.ReadFile(countFile)
	if os.IsNotExist(err) {
		return 0
	}
	if err != nil {
		t.Fatal(err)
	}
	return len(data) / 2 // "x\n" per invocation
}

func TestCliCredentialsMissingBinary(t *testing.T) {
	credentials := &cliCredentials{command: filepath.Join(t.TempDir(), "no-such-binary")}
	_, err := credentials.Token(context.Background())
	if !errors.Is(err, ErrCliCredentials) {
		t.Errorf("missing binary must map to ErrCliCredentials, got %v", err)
	}
}

func TestCliCredentialsInvalidJSON(t *testing.T) {
	command, _ := fakeCli(t, `echo "not json"`)
	credentials := &cliCredentials{command: command}
	_, err := credentials.Token(context.Background())
	if !errors.Is(err, ErrCliCredentials) {
		t.Errorf("invalid JSON must map to ErrCliCredentials, got %v", err)
	}
}

func TestCliCredentialsNonZeroExit(t *testing.T) {
	command, _ := fakeCli(t, `echo "run: whodb login" >&2; exit 1`)
	credentials := &cliCredentials{command: command}
	_, err := credentials.Token(context.Background())
	if !errors.Is(err, ErrCliCredentials) {
		t.Errorf("CLI failure must map to ErrCliCredentials, got %v", err)
	}
}

func TestCliCredentialsCachesUntilNearExpiry(t *testing.T) {
	expiry := time.Now().Add(1 * time.Hour).Format(time.RFC3339)
	command, countFile := fakeCli(t, fmt.Sprintf(
		`echo '{"access_token":"tok-1","expires_at":"%s"}'`, expiry))
	credentials := &cliCredentials{command: command}
	for range 3 {
		token, err := credentials.Token(context.Background())
		if err != nil {
			t.Fatal(err)
		}
		if token != "tok-1" {
			t.Fatalf("token: got %q", token)
		}
	}
	if got := execCount(t, countFile); got != 1 {
		t.Errorf("a fresh token must be served from cache: %d execs", got)
	}
}

func TestCliCredentialsReExecsNearExpiry(t *testing.T) {
	expiry := time.Now().Add(30 * time.Second).Format(time.RFC3339) // inside the 60s skew
	command, countFile := fakeCli(t, fmt.Sprintf(
		`echo '{"access_token":"tok-1","expires_at":"%s"}'`, expiry))
	credentials := &cliCredentials{command: command}
	for range 2 {
		if _, err := credentials.Token(context.Background()); err != nil {
			t.Fatal(err)
		}
	}
	if got := execCount(t, countFile); got != 2 {
		t.Errorf("a near-expiry token must re-exec every call: %d execs", got)
	}
}

func TestCliCredentialsRefreshDropsCache(t *testing.T) {
	expiry := time.Now().Add(1 * time.Hour).Format(time.RFC3339)
	command, countFile := fakeCli(t, fmt.Sprintf(
		`echo '{"access_token":"tok-1","expires_at":"%s"}'`, expiry))
	credentials := &cliCredentials{command: command}
	if _, err := credentials.Token(context.Background()); err != nil {
		t.Fatal(err)
	}
	credentials.Refresh()
	if _, err := credentials.Token(context.Background()); err != nil {
		t.Fatal(err)
	}
	if got := execCount(t, countFile); got != 2 {
		t.Errorf("Refresh must force a re-exec: %d execs", got)
	}
}

File: sdk/packages/typescript/test/auth.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, writeFileSync, chmodSync, readFileSync, existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { cliProvider } from '../dist/auth.js';
import { CliCredentialsError } from '../dist/errors.js';

/** Writes a fake whodb CLI script; appends to a counter file per invocation. */
function fakeCli(body) {
  const dir = mkdtempSync(join(tmpdir(), 'whodb-cli-test-'));
  const countFile = join(dir, 'count');
  const command = join(dir, 'whodb');
  writeFileSync(command, `#!/bin/sh\necho x >> "${countFile}"\n${body}\n`);
  chmodSync(command, 0o755);
  return { command, execCount: () => (existsSync(countFile) ? readFileSync(countFile, 'utf8').length / 2 : 0) };
}

test('missing CLI binary maps to CliCredentialsError', async () => {
  const provider = cliProvider(join(tmpdir(), 'no-such-whodb-binary'));
  await assert.rejects(() => provider.token(), CliCredentialsError);
});

test('invalid CLI JSON maps to CliCredentialsError', async () => {
  const { command } = fakeCli('echo "not json"');
  await assert.rejects(() => cliProvider(command).token(), CliCredentialsError);
});

test('CLI failure surfaces as CliCredentialsError', async () => {
  const { command } = fakeCli('echo "run: whodb login" >&2; exit 1');
  await assert.rejects(() => cliProvider(command).token(), CliCredentialsError);
});

test('fresh tokens are served from cache without re-exec', async () => {
  const expiry = new Date(Date.now() + 3600_000).toISOString();
  const { command, execCount } = fakeCli(
    `echo '{"access_token":"tok-1","expires_at":"${expiry}"}'`,
  );
  const provider = cliProvider(command);
  for (let i = 0; i < 3; i += 1) {
    assert.equal(await provider.token(), 'tok-1');
  }
  assert.equal(execCount(), 1);
});

test('near-expiry tokens re-exec the CLI', async () => {
  const expiry = new Date(Date.now() + 30_000).toISOString(); // inside the 60s skew
  const { command, execCount } = fakeCli(
    `echo '{"access_token":"tok-1","expires_at":"${expiry}"}'`,
  );
  const provider = cliProvider(command);
  await provider.token();
  await provider.token();
  assert.equal(execCount(), 2);
});

test('refresh drops the cache and re-execs', async () => {
  const expiry = new Date(Date.now() + 3600_000).toISOString();
  const { command, execCount } = fakeCli(
    `echo '{"access_token":"tok-1","expires_at":"${expiry}"}'`,
  );
  const provider = cliProvider(command);
  await provider.token();
  await provider.refresh();
  await provider.token();
  assert.equal(execCount(), 2);
});
